### ISSUE TITLE
Variable outputs of stochastic functions should never require grad

### ISSUE THREAD

**Issue by apaszke** [pytorch/pytorch#4159]
I don't think we should bother to implement reparametrized gradients for Variable functions (`torch.distributions` will have them + I don't think it's possible for bernoulli), and if we keep setting these outputs to `requires_grad=True` and pushing back zeros some people might be confused. Also, this doesn't match the legacy behavior, and can make some in-place checks that passed before fail.

Let's just make them non-differentiable...

**Comment by fritzo** [comment]
Sounds reasonable to me. Note reprameterization is [not uniquely defined](tasks) for multivariate samplers (e.g. MultivariateNormal) or for circular samplers (e.g. VonMises), so it is nice to separate concepts between low-level random generation and high-level pathwise differentiable stochastic functions.
